workflow-manager/main.go

```diff
@@ -254,7 +389,270 @@ func getReadyBatchesGS(ctx context.Context, inputBucket string) ([]string, error
+			b.sig = true
+		}
+	}
+
+	var readyBatches []string
+	for k, v := range batches {
+		if v.metadata && v.avro && v.sig {
+			log.Printf("ready: %s", k)
+			readyBatches = append(readyBatches, k)
+		} else {
+			log.Printf("unready: %s", k)
+		}
+	}
+
+	return readyBatches, nil
+}
+
+func getReadyValidationBatchesS3(ctx context.Context, validationsBucket string, isFirst bool, filter []string, roleARN string) ([]string, error) {
+	parts := strings.SplitN(validationsBucket, "/", 2)
+	region := parts[0]
+	bucket := parts[1]
+	sess, err := aws_session.NewSession()
+	if err != nil {
+		return nil, fmt.Errorf("making AWS session: %w", err)
+	}
+
+	stsSTS := sts.New(sess)
+	roleSessionName := ""
+	roleProvider := stscreds.NewWebIdentityRoleProviderWithToken(
+		stsSTS, roleARN, roleSessionName, tokenFetcher{})
+
+	credentials := credentials.NewCredentials(roleProvider)
+
+	config := aws.NewConfig().
+		WithRegion(region).
+		WithCredentials(credentials)
+	svc := s3.New(sess, config)
+	resp, err := svc.ListObjectsV2(&s3.ListObjectsV2Input{Bucket: aws.String(bucket)})
+	if err != nil {
+		return nil, fmt.Errorf("Unable to list items in bucket %q, %w", validationsBucket, err)
+	}
+	batches := make(map[string]*batch)
+	for _, item := range resp.Contents {
+		name := *item.Key
+		basename := validationBasename(name, isFirst)
+		if !contains(filter, basename) {
+			log.Printf("ignoring batch %s", basename)
+			continue
+		}
+		if batches[basename] == nil {
+			batches[basename] = new(batch)
+		}
+		b := batches[basename]
+		if strings.HasSuffix(name, fmt.Sprintf(".validity_%s", index(isFirst))) {
+			b.metadata = true
+		}
+		if strings.HasSuffix(name, fmt.Sprintf(".validity_%s.avro", index(isFirst))) {
+			b.avro = true
+		}
+		if strings.HasSuffix(name, fmt.Sprintf(".validity_%s.sig", index(isFirst))) {
+			b.sig = true
+		}
+	}
+
+	var readyBatches []string
+	for k, v := range batches {
+		if v.metadata && v.avro && v.sig {
+			log.Printf("ready: %s", k)
+			readyBatches = append(readyBatches, k)
+		} else {
+			log.Printf("unready: %s", k)
+		}
+	}
+	return readyBatches, nil
+}
+
+func launchAggregationJob(ctx context.Context, readyBatches []string) error {
+	if len(readyBatches) == 0 {
+		log.Printf("no batches to aggregate")
+		return nil
+	}
+
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return fmt.Errorf("cluster config: %w", err)
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return fmt.Errorf("clientset: %w", err)
+	}
+	aggregationID := ""
+	var earliestBatch *batchPath
+	var latestBatch *batchPath
+	batchIDs := []string{}
+	batchTimes := []string{}
+
+	for _, batchName := range readyBatches {
+		batchPath, err := newBatchPath(batchName)
+		if err != nil {
+			return err
+		}
+		batchIDs = append(batchIDs, batchPath.batchID)
+		batchTimes = append(batchTimes, batchPath.batchDateString())
+
+		// All batches should have the same aggregation ID?
```

Comment:
This is an error that could be triggered by a mistake our peer makes, and it would bring processing to a halt until they resolved it. We should be robust to the error (e.g. log an error and ignore the mismatch), or enforce this property internally by doing a filter on the readyBatches before passing them to this function.

Reply:
I'm honestly not sure if the aggregation ID ever changes, but wanted to get any aggregation working before circling back to that question, or implementing logic to enforce it.